Allowing readable verification of instrumentation results

jcouv opened the pull request:
**Customer scenario**

This is a test-only change, with no customer impact. LUT has some tests that are difficult to review.
This PR offers a small test framework which will make such tests more readable for code reviews.

**Bugs this fixes:** 

Fixes https://github.com/dotnet/roslyn/issues/15818

**Risk**
**Performance impact**

None. Test only.

@dotnet/roslyn-compiler @JohnHamby @tmat for review.

JohnHamby commented on src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicInstrumentationTests.cs:
```diff
@@ -2120,5 +2118,159 @@ private CompilationVerifier CompileAndVerify(string source, string expectedOutpu
+                    var actualSnippets = GetActualSnippets(method, reader, sourceLines);
+                    var expectedSnippets = spanExpectations[method].SnippetExpectations;
+
+                    if (actualSnippets.Length != expectedSnippets.Length)
```
This if statement could be replaced with "Assert.AreEqual(...)".

JohnHamby commented on src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicInstrumentationTests.cs:
```diff
@@ -2120,5 +2118,159 @@ private CompilationVerifier CompileAndVerify(string source, string expectedOutpu
+                        var expectedSnippet = expectedSnippets[index];
+                        if (!string.IsNullOrWhiteSpace(expectedSnippet))
+                        {
+                            if (!actualSnippet.StartsWith(expectedSnippet))
```
This if statement could be replaced with "Assert.True(actualSnippet.StartsWith(....".

JohnHamby commented on src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicInstrumentationTests.cs:
```diff
@@ -2120,5 +2118,159 @@ private CompilationVerifier CompileAndVerify(string source, string expectedOutpu
+                        }
+                        else
+                        {
+                            Assert.True(false, $"The {index}-th expected snippet for method {method} was missing. Actual: {actualSnippet}");
```
Is there an Assert.Fail, or some such?

JohnHamby commented:
I can see how one might find this form easier to follow, so I endorse this.

jaredpar commented on src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicInstrumentationTests.cs:
```diff
@@ -4,6 +4,11 @@
 using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
 using Microsoft.CodeAnalysis.Emit;
 using Xunit;
+using Microsoft.CodeAnalysis.Test.Utilities;
```
Nit: sort usings

jaredpar commented on src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicInstrumentationTests.cs:
```diff
@@ -2120,5 +2118,159 @@ private CompilationVerifier CompileAndVerify(string source, string expectedOutpu
 
         private static readonly MetadataReference[] s_refs = new[] { MscorlibRef_v4_0_30316_17626, SystemRef_v4_0_30319_17929, SystemCoreRef_v4_0_30319_17929, ValueTupleRef, SystemRuntimeFacadeRef };
+
+        class InstrumentationChecker
```
Use explicit visibility modifiers

jaredpar commented on src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicInstrumentationTests.cs:
```diff
@@ -2120,5 +2118,159 @@ private CompilationVerifier CompileAndVerify(string source, string expectedOutpu
+        class InstrumentationChecker
+        {
+            StringBuilder consoleExpectations = new StringBuilder();
+            Dictionary<int /* method */, MethodChecker> spanExpectations = new Dictionary<int, MethodChecker>();
```
Use explicit visibility modifiers, prefix private fields with `_`

tmat commented on src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicInstrumentationTests.cs:
```diff
@@ -2120,5 +2118,159 @@ private CompilationVerifier CompileAndVerify(string source, string expectedOutpu
+            /// <summary>
+            /// Start recording expectations for a method, but without checking snippets for this method.
+            /// </summary>
+            public MethodChecker M(int method, int file)
```
I'd just call this ```Method```. What are the numbers?

tmat replied:
Please add a comment to the source

tmat commented on src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicInstrumentationTests.cs:
```diff
@@ -2120,5 +2118,159 @@ private CompilationVerifier CompileAndVerify(string source, string expectedOutpu
+                        Assert.False(true, $"The number of spans was expected to be {expectedSnippets.Length} for the method {method}, but it actually has {actualSnippets.Length}.");
+                    }
+
+                    for (int index = 0; index < actualSnippets.Length; index++)
```
Can we use ```AssertEx.Equal(IEnumerable, IEnumerable, ...)```, so that if the test fails it prints out the expected test source?

jcouv replied:
More generally on the usability question, I've only had limited experience using the APIs and responding to the assert failures (only ported one test). But so far, my impression is that it provided the right guidance. 
This should be no problem to tweak once we get more hands-on experience.

tmat replied:
The scenarios: the initial test source and when something breaks the test show before and after and a diff.

```AssertEx.Equal``` is pretty powerful -- it implements basic line-by-line differ. 
http://source.roslyn.io/#System.Reflection.Metadata.Tests/TestUtilities/AssertEx.cs,95a4e29ac2ebf80c

It's used for IL and PDB tests. The idea is basically that when comparing sequences of things it does the comparison and if is not equal it prints out results (with a diff) using a custom printer. So you can craft it to print out the actual and expected source code of the test that can be copy-pasted to the actual source.

jcouv replied:
My last change uses `AssertEx` for sequence comparison. The output is indeed more informative :-)
Below is an example. Thanks!

```
Validation of method 1 failed.

Expected:
Public Sub Main,
WRONG,
Microsoft.CodeAnalysis.Runtime.Instrumentation.FlushPayload()
Actual:
Public Sub Main(args As String())                                   ' Method 1,
TestMain(),
Microsoft.CodeAnalysis.Runtime.Instrumentation.FlushPayload()
Differences:
    Public Sub Main(args As String())                                   ' Method 1,
++> TestMain(),
--> WRONG,
    Microsoft.CodeAnalysis.Runtime.Instrumentation.FlushPayload()
```

jcouv commented:
Apologies for snowballing the PR. I added support for VB tests, which was a bit more involved than I expected.
I'm still looking into Tomasz's suggestion to use `AssertEx`'s sequence comparisons.

AdamSpeight2008 commented on src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicInstrumentationTests.cs:
```diff
@@ -707,50 +708,43 @@ public class Program
+
+            var checker = new CSharpInstrumentationChecker();
+            checker.Method(3, 1, "public int Prop3")
+                .T("get");
```
Speaking of readability, it took some time to realise `T` and `F` stand for True and False. Could we use True and False instead.

AdamSpeight2008 commented on src/Test/Utilities/Desktop/InstrumentationChecker.cs:
```diff
@@ -0,0 +1,267 @@
+
+namespace Microsoft.CodeAnalysis.Test.Utilities
+{
+    internal sealed class CSharpInstrumentationChecker : BaseInstrumentationChecker
```
Could you add some doc comments, to these methods. To help those that are external or new to the project use them.

AdamSpeight2008 commented on src/Test/Utilities/Desktop/InstrumentationChecker.cs:
```diff
@@ -0,0 +1,267 @@
+
+        internal override void Dump(DynamicAnalysisDataReader reader, string[] sourceLines)
+        {
+            var output = new StringBuilder();
```
Maybe a better to use a pooled version.

jcouv commented:
From discussion with Chuck, I will try to see if the `Dump` method could be eliminated (and have `AssertEx` do all the formatting work to produce a friendly output).

jcouv commented:
@cston I got a solution working that leverages `AssertEx`'s formatting, but I'm afraid the experience was worse. 
Basically, the problem is that the output of the test cannot be what the user should type in his code (because we don't have access to the actual flags from the compilation's execution). So the output from `AssertEx` feels quite misleading.

I think we should leave the `Dump` method as-is for now. 

If that makes sense, this PR is ready for review. 
Note that because of conflicts with `master`, I rebased the whole thing and squashed in the process.

jcouv commented:
@JohnHamby @tmat for review please.

JohnHamby commented:
Certainly I'm fine with this. This was a lot of clever work, and it does make the tests easier to verify by eye.

cston commented on src/Test/Utilities/Desktop/InstrumentationChecker.cs:
```diff
@@ -0,0 +1,398 @@
+                else
+                {
+                    Assert.True(expectedSourceSnippet == null,
+                        "You must pass a snippet when checking the method with M if you intend to verify snippets within the method.");
```
Is the assert text correct? If the assert fails, a (non-null) snippet was passed in. Also, should `M` be `Method`?

jcouv replied:
Sorry, I'd missed that comment. We can fix opportunistically next time someone writes LUT tests with this helper.

cston commented:
LGTM

jcouv commented:
Thanks. I'll go ahead and merge given this is a test-only change.
